File: jdroid-android-firebase-fcm/src/main/java/com/jdroid/android/firebase/fcm/FcmRegistrationWorker.java
package com.jdroid.android.firebase.fcm;

import android.content.Context;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.InstanceIdResult;
import com.jdroid.android.application.AbstractApplication;
import com.jdroid.android.google.GooglePlayServicesUtils;
import com.jdroid.android.jetpack.work.AbstractWorker;
import com.jdroid.java.exception.UnexpectedException;
import com.jdroid.java.utils.LoggerUtils;

import org.slf4j.Logger;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import androidx.work.Constraints;
import androidx.work.Data;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.ExistingWorkPolicy;
import androidx.work.NetworkType;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;
import androidx.work.WorkerParameters;

public class FcmRegistrationWorker extends AbstractWorker {
	
	public static final String WORK_MANAGER_TAG = "fcm";
	
	private final static Logger LOGGER = LoggerUtils.getLogger(FcmRegistrationWorker.class);

	private final static String UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA = "updateLastActiveTimestamp";

	public FcmRegistrationWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
		super(context, workerParams);
	}

	public static void start(Boolean updateLastActiveTimestamp) {
		OneTimeWorkRequest.Builder requestBuilder = new OneTimeWorkRequest.Builder(FcmRegistrationWorker.class);
		
		Data.Builder dataBuilder = new Data.Builder();
		dataBuilder.putBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, updateLastActiveTimestamp);
		requestBuilder.setInputData(dataBuilder.build());
		
		Constraints.Builder constrainsBuilder = new Constraints.Builder();
		constrainsBuilder.setRequiredNetworkType(NetworkType.CONNECTED);
		requestBuilder.setConstraints(constrainsBuilder.build());
		
		requestBuilder.addTag(WORK_MANAGER_TAG);
		
		WorkManager.getInstance().beginUniqueWork(FcmRegistrationWorker.class.getSimpleName(), ExistingWorkPolicy.KEEP, requestBuilder.build()).enqueue();
	}
	
	// By Google recommendation, we should execute this worker every 2 weeks, to have always fresh tokens on server side
	private void startPeriodic(Boolean updateLastActiveTimestamp) {
		PeriodicWorkRequest.Builder requestBuilder = new PeriodicWorkRequest.Builder(FcmRegistrationWorker.class, 14, TimeUnit.DAYS, 7, TimeUnit.DAYS);
		
		Data.Builder dataBuilder = new Data.Builder();
		dataBuilder.putBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, updateLastActiveTimestamp);
		requestBuilder.setInputData(dataBuilder.build());
		
		Constraints.Builder constrainsBuilder = new Constraints.Builder();
		constrainsBuilder.setRequiredNetworkType(NetworkType.CONNECTED);
		requestBuilder.setConstraints(constrainsBuilder.build());
		
		requestBuilder.addTag(WORK_MANAGER_TAG);
		
		WorkManager.getInstance().enqueueUniquePeriodicWork(FcmRegistrationWorker.class.getSimpleName(), ExistingPeriodicWorkPolicy.KEEP, requestBuilder.build());
	}
	
	@NonNull
	@Override
	protected Result onWork() {
		if (GooglePlayServicesUtils.isGooglePlayServicesAvailable(getApplicationContext())) {
			Boolean updateLastActiveTimestamp = getInputData().getBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, false);
			for (FcmSender fcmSender : AbstractFcmAppModule.get().getFcmSenders()) {
				String registrationToken;
				try {
					registrationToken = getRegistrationToken(fcmSender.getSenderId());
				} catch (Exception e) {
					AbstractApplication.get().getExceptionHandler().logHandledException("Error when getting FCM registration token. Will retry later.", e);
					return Result.RETRY;
				}
				
				if (registrationToken == null) {
					LOGGER.info("Null registration token. Will retry later.");
					return Result.RETRY;
				}

				try {
					LOGGER.info("Registering FCM token on server");
					fcmSender.onRegisterOnServer(registrationToken, updateLastActiveTimestamp, getInputData().getKeyValueMap());
				} catch (Exception e) {
					AbstractApplication.get().getExceptionHandler().logHandledException("Failed to register the device on server. Will retry later.", e);
					return Result.RETRY;
				}
			}
			startPeriodic(updateLastActiveTimestamp);
			return Result.SUCCESS;
		} else {
			LOGGER.warn("FCM not initialized because Google Play Services is not available");
			return Result.RETRY;
		}
	}

	@WorkerThread
	@Nullable
	public static String getRegistrationToken(String senderId) {
		if (GooglePlayServicesUtils.isGooglePlayServicesAvailable(AbstractApplication.get())) {
			String registrationToken;
			try {
				if (senderId != null) {
					registrationToken = FirebaseInstanceId.getInstance().getToken(senderId, "FCM");
					LOGGER.info("Registration token for sender id [" + senderId + "]: " + registrationToken);
				} else {
					Task<InstanceIdResult> task = FirebaseInstanceId.getInstance().getInstanceId();
					// Block on the task and get the result synchronously
					InstanceIdResult instanceIdResult = Tasks.await(task);
					registrationToken = instanceIdResult.getToken();
					LOGGER.info("Registration token for default sender id: " + registrationToken);
				}
			} catch (IOException | ExecutionException | InterruptedException e) {
				throw new UnexpectedException(e);
			}
			return registrationToken;
		}
		return null;
	}
}

File: jdroid-android-firebase-fcm/src/main/java/com/jdroid/android/firebase/fcm/FcmRegistrationWorker.kt
package com.jdroid.android.firebase.fcm

import android.content.Context
import androidx.annotation.WorkerThread
import androidx.work.Constraints
import androidx.work.Data
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.ExistingWorkPolicy
import androidx.work.NetworkType
import androidx.work.OneTimeWorkRequest
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import com.google.android.gms.tasks.Tasks
import com.google.firebase.iid.FirebaseInstanceId
import com.jdroid.android.application.AbstractApplication
import com.jdroid.android.firebase.fcm.AbstractFcmAppModule.Companion.get
import com.jdroid.android.google.GooglePlayServicesUtils.isGooglePlayServicesAvailable
import com.jdroid.android.jetpack.work.AbstractWorker
import com.jdroid.java.exception.UnexpectedException
import com.jdroid.java.utils.LoggerUtils
import java.io.IOException
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

class FcmRegistrationWorker(context: Context, workerParams: WorkerParameters) : AbstractWorker(context, workerParams) {

    companion object {

        const val WORK_MANAGER_TAG = "fcm"
        private val LOGGER = LoggerUtils.getLogger(FcmRegistrationWorker::class.java)
        private const val UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA = "updateLastActiveTimestamp"

        fun enqueue(updateLastActiveTimestamp: Boolean) {
            val requestBuilder = OneTimeWorkRequest.Builder(FcmRegistrationWorker::class.java)
            val dataBuilder = Data.Builder()
            dataBuilder.putBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, updateLastActiveTimestamp)
            requestBuilder.setInputData(dataBuilder.build())
            val constrainsBuilder = Constraints.Builder()
            constrainsBuilder.setRequiredNetworkType(NetworkType.CONNECTED)
            requestBuilder.setConstraints(constrainsBuilder.build())
            requestBuilder.addTag(WORK_MANAGER_TAG)
            WorkManager.getInstance(AbstractApplication.get())
                .beginUniqueWork(FcmRegistrationWorker::class.java.simpleName, ExistingWorkPolicy.KEEP, requestBuilder.build()).enqueue()
        }

        @WorkerThread
        fun getRegistrationToken(senderId: String?): String? {
            if (isGooglePlayServicesAvailable(AbstractApplication.get())) {
                val registrationToken: String?
                try {
                    if (senderId != null) {
                        registrationToken = FirebaseInstanceId.getInstance().getToken(senderId, "FCM")
                        LOGGER.info("Registration token for sender id [$senderId]: $registrationToken")
                    } else {
                        val task = FirebaseInstanceId.getInstance().instanceId
                        // Block on the task and get the result synchronously
                        val instanceIdResult = Tasks.await(task)
                        registrationToken = instanceIdResult.token
                        LOGGER.info("Registration token for default sender id: $registrationToken")
                    }
                } catch (e: IOException) {
                    throw UnexpectedException(e)
                } catch (e: ExecutionException) {
                    throw UnexpectedException(e)
                } catch (e: InterruptedException) {
                    throw UnexpectedException(e)
                }
                return registrationToken
            }
            return null
        }
    }

    override fun onWork(): Result {
        return if (isGooglePlayServicesAvailable(applicationContext)) {
            val updateLastActiveTimestamp = inputData.getBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, false)
            for (fcmSender in AbstractFcmAppModule.get().getFcmSenders()) {
                val registrationToken: String?
                try {
                    registrationToken = getRegistrationToken(fcmSender.getSenderId())
                } catch (e: Exception) {
                    AbstractApplication.get().exceptionHandler.logHandledException("Error when getting FCM registration token. Will retry later.", e)
                    return Result.retry()
                }

                if (registrationToken == null) {
                    LOGGER.info("Null registration token. Will retry later.")
                    return Result.retry()
                }

                try {
                    LOGGER.info("Registering FCM token on server")
                    fcmSender.onRegisterOnServer(registrationToken, updateLastActiveTimestamp, inputData.keyValueMap)
                } catch (e: Exception) {
                    AbstractApplication.get().exceptionHandler.logHandledException("Failed to register the device on server. Will retry later.", e)
                    return Result.retry()
                }
            }
            startPeriodic(updateLastActiveTimestamp)
            Result.success()
        } else {
            LOGGER.warn("FCM not initialized because Google Play Services is not available")
            Result.retry()
        }
    }

    // By Google recommendation, we should execute this worker every 2 weeks, to have always fresh tokens on server side
    private fun startPeriodic(updateLastActiveTimestamp: Boolean) {
        val requestBuilder = PeriodicWorkRequest.Builder(FcmRegistrationWorker::class.java,
            14L, TimeUnit.DAYS,
            7L, TimeUnit.DAYS
        )
        val dataBuilder = Data.Builder()
        dataBuilder.putBoolean(UPDATE_LAST_ACTIVE_TIMESTAMP_EXTRA, updateLastActiveTimestamp)
        requestBuilder.setInputData(dataBuilder.build())
        val constrainsBuilder = Constraints.Builder()
        constrainsBuilder.setRequiredNetworkType(NetworkType.CONNECTED)
        requestBuilder.setConstraints(constrainsBuilder.build())
        requestBuilder.addTag(WORK_MANAGER_TAG)
        WorkManager.getInstance(AbstractApplication.get()).enqueueUniquePeriodicWork(
            FcmRegistrationWorker::class.java.simpleName,
            ExistingPeriodicWorkPolicy.KEEP,
            requestBuilder.build()
        )
    }
}
